sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -2011,18 +2040,37 @@ def task_worker(self, stopping_event, sfp_error_event):
 
         helper_logger.log_info("Stop SFP monitoring loop")
 
-    def task_run(self, sfp_error_event):
+    def run(self):
+        self.thread_id = threading.current_thread().ident
         if self.task_stopping_event.is_set():
             return
-
-
-        self.task_process = multiprocessing.Process(target=self.task_worker, args=(
-            self.task_stopping_event, sfp_error_event))
-        self.task_process.start()
-
-    def task_stop(self):
+        try:
+            self.task_worker(self.task_stopping_event, self.sfp_error_event)
+        except Exception as e:
+            helper_logger.log_error("Exception occured at {} thread due to {}".format(threading.current_thread().getName(), repr(e)))
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+            msg = traceback.format_exception(exc_type, exc_value, exc_traceback)
+            for tb_line in msg:
+                for tb_line_split in tb_line.splitlines():
+                    helper_logger.log_error(tb_line_split)
+            self.exc = e
+            self.main_thread_stop_event.set()
+
+    # SfpStateUpdateTask thread has a call to an API which could potentially sleep in the order of seconds and hence,
+    # could block the xcvrd daemon graceful shutdown process for a prolonged time. Raising an exception will allow us to
+    # interrupt the SfpStateUpdateTask thread while sleeping and will allow graceful shutdown of the thread
+    def raise_exception(self):
+        res = ctypes.pythonapi.PyThreadState_SetAsyncExc(ctypes.c_ulong(self.thread_id),
+              ctypes.py_object(SystemExit))
+        if res > 1:
+            ctypes.pythonapi.PyThreadState_SetAsyncExc(ctypes.c_ulong(self.thread_id), 0)
+            helper_logger.log_error('Exception raise failure for SfpStateUpdateTask')
+
+    def join(self):
```

Comment:
it seems these 3 newly added "join" functions are all the same, suggest finding someway to reduce duplicated code.

Reply:
Based on further changes, I have modified join function of CmisManagerTask to ensure that threading.Thread.join() is not called if skip_cmis_mgr is set. This is required to ensure that on Mellanox wherein CMIS thread is not spawned, xcvrd doesn't end up calling CmisManagerTask.join() and eventually results in a infinite wait and hence, prompting supervisord to kill xcvrd forcefully during graceful shutdown.
This leaves 2 other places with duplicated code for "join" function which I have planned to leave it currently as is to allow future modifications if required.